File: PythonCode/InfoAnalysis.py
import numpy as np
import pylab as plt
import pickle
import os;
#from Parameters import *


class InfoAnalysis(object):
    def loadParams(self,borrowed_globals):
        globals().update(borrowed_globals);
    
    
    def singleCellInfoAnalysis(self,experimentName,phases,saveImage = True, showImage = True, nBins=3,weightedAnalysis = False,plotAllSingleCellInfo = True):
        fig=plt.figure(4 , figsize=(20, 5),dpi=150);
        
#         plotFRMap = True;
#         Parameters

        nExcitCells = exDim*exDim;#32*32;
        nInhibCells = inDim*inDim;#16*16;
        
        useMaxFRTh = True;
        
        maxFRTh = 100;
#         
#         dimExcitLayers = 64;
#         
# 
#         nObj = 3;
#         nTrans = 2;
#         nLayers = 4;
#         presentationTime = 3.0;
        

        nInfoCalc = nExcitCells;
        
        phaseIndex = 1;
        for phase in phases:

#             fn_id = "../output/Neurons_SpikeIDs_" + phase + "_Epoch0.txt";
#             fn_t = "../output/Neurons_SpikeTimes_" + phase + "_Epoch0.txt";
#             
#             spikeIDs = np.loadtxt(fn_id);
#             spikeTimes = np.loadtxt(fn_t);
            
            fn_id = "../output/"+experimentName+"/Neurons_SpikeIDs_" + phase + "_Epoch0.bin";
            fn_t = "../output/"+experimentName+"/Neurons_SpikeTimes_" + phase + "_Epoch0.bin";  
            dtIDs = np.dtype('int32');
            dtTimes = np.dtype('f4');
            
            spikeIDs = np.fromfile(fn_id, dtype=dtIDs);
            spikeTimes = np.fromfile(fn_t, dtype=dtTimes);
            

            FR = np.zeros((nObj, nTrans,nLayers, nExcitCells));
            for l in range(0,nLayers):
                cond_ids = (l*(nExcitCells+nInhibCells) < spikeIDs) & (spikeIDs < l*(nInhibCells+nExcitCells)+nExcitCells);
                spikeIDs_layer = np.extract(cond_ids, spikeIDs);
                spikeTimes_layer = np.extract(cond_ids, spikeTimes);
                for obj in range(nObj):
                    for trans in range(nTrans):
                        cond_stim = ((presentationTime*(obj*nTrans+trans)) < spikeTimes_layer) & (spikeTimes_layer < (presentationTime*(obj*nTrans+trans+1)));
                        spikeIDs_stim = np.extract(cond_stim,spikeIDs_layer)-((nExcitCells+nInhibCells)*l);
                        for id in spikeIDs_stim:
                            FR[obj,trans,l,id]=FR[obj,trans,l,id]+1;
#             FR = np.random.rand(nObj, nTrans,nLayers, nExcitCells)
        
        
#             if(plotFRMap):
#                 for obj in range(nObj):
#                     for l in range(nLayers):
#                         for trans in range(nTrans):
#                             plt.subplot(nLayers, nTrans*nObj, (nLayers-l-1)*(nTrans*nObj)+(obj*nTrans)+trans+1);
#                             plt.title('Firing Rate Map: obj ' + str(obj) )
#                             
#                             FRMap = np.zeros((exDim,exDim));
#                             for y in range(exDim):
#                                 for x in range(exDim):
#                                     id = x*exDim + y;
#                                     FRMap[y,x] = FR[obj,trans,l,id];
#                             Rmax = 200;
#                             plt.imshow(FRMap, cmap='jet', interpolation='none', vmin=0, vmax=Rmax)
#                             plt.colorbar();
#                 plt.show();
#                 fig.savefig("../output/AvgFR.png");
            
#             print FR;
                        
            performanceMeasure = 0.0;
            FR/=presentationTime;
            
            for l in range(0,nLayers):
                if(not useMaxFRTh):
                    maxFRTh = FR[:,:,l,:].max()
                print(" Maximum Firing Rate Threshold of " + str(maxFRTh) +" is used");
                
                if FR[:,:,l,:].max()>0.001:
                    FR_tmp = FR/maxFRTh;
                else:
                    FR_tmp = FR;
                    
                    
                infos = np.zeros(nExcitCells);
                
                sumPerBin = np.zeros((nExcitCells,nBins));
                sumPerObj = nTrans;
                sumPerCell = nTrans*nObj;
                IRs = np.zeros((nObj,nExcitCells));#I(R,s) single cell information
                IRs_weighted = np.zeros((nObj,nExcitCells));#I(R,s) single cell information
                pq_r = np.zeros(nObj)#prob(s') temporally used in the decoing process
                Ps = 1/nObj   #Prob(s) 
                

                
                print("**Loading data**")
                binMatrix = np.zeros((nExcitCells, nObj, nBins));# #number of times when fr is classified into a specific bin within a specific objs's transformations
                for obj in range(nObj):
                    print str(obj) + '/' + str(nObj);
                    for trans in range(nTrans):
                        for cell in range(nExcitCells):
#                             bin = np.around(FR_tmp[obj,trans,l,cell]*(nBins-1));
                            bin = min(np.floor((FR_tmp[obj,trans,l,cell])*(nBins)),nBins-1)
                            binMatrix[cell,obj,bin]=binMatrix[cell,obj,bin]+1;
#                             
#                             
                
                
                #print binMatrix;
#                 for obj in range(nObj):
#                     print str(obj) + '/' + str(nObj);
#                     for trans in range(nTrans):
#                         for cell in range(nExcitCells):                            
# 
#                             hist = np.histogram(FR[obj,:,l,cell], bins=range(nBins+1), range=(0,maxFRTh));
#                             for bin in range(nBins):
#                                 binMatrix[cell,obj,bin]=hist[0][bin];
                                    
                
                
                print "** single-cell information analysis **";
                # Loop through all cells to calculate single cell information
                for cell in range(nExcitCells):
                    # For each cell, count the number of transforms per bin
                    for bin in range(nBins):
                        for obj in range(nObj):
                            sumPerBin[cell,bin]+=binMatrix[cell,obj,bin];
            
                    # Calculate the information for cell_x cell_y per stimulus
                    for obj in range(nObj):
                        for bin in range(nBins):
                            Pr = sumPerBin[cell,bin]/sumPerCell;
                            Prs = binMatrix[cell,obj,bin]/sumPerObj;
                            if(Pr!=0 and Prs!=0 and Pr<Prs):
                                IRs[obj,cell]+=(Prs*(np.log2(Prs/Pr)));#*((bin-1)/(nBins-1)); #could be added to weight the degree of firing rates.
                                #IRs(row,col,obj)=IRs(row,col,obj)+(Prs*(log2(Prs/Pr)))*((bin-1)/(nBins-1)); #could be added to weight the degree of firing rates.
                                IRs_weighted[obj,cell]+=(Prs*(np.log2(Prs/Pr)))*((bin)/(nBins-1)); #could be added to weight the degree of firing rates.
             
                if (weightedAnalysis):
                    IRs = IRs_weighted;
                
                IRs_sorted = np.sort(IRs*-1)*-1;
                np.savetxt("../output/"+experimentName+"/SingleCellInfo_l" + str(l) + phase + ".csv",IRs_sorted, delimiter=',');

                
                
                if (plotAllSingleCellInfo):
                    IRs_sorted = np.sort(IRs*-1)*-1;
                    plt.subplot(2,nLayers,(phaseIndex-1)*nLayers+(l+1));

#                     plt.plot(np.transpose(IRs_sorted), color='k');
                    plt.plot(np.transpose(IRs_sorted));
                    
                    plt.ylim([-0.05, np.log2(nObj)+0.05]);
                    plt.xlim([0, nExcitCells])
                    plt.title("Layer " + str(l));
                    plt.ylabel("Information [bit]");
                    plt.xlabel("Cell Rank");
                    
                else:
                    IRs = np.max(IRs,axis=0);
                    
                    IRs_sorted = np.transpose(np.sort(IRs));
                    reversed_arr = IRs_sorted[::-1]
                
                
                    infos = reversed_arr;
                    np.savetxt("../output/"+experimentName+"/SingleCellInfo_l" + str(l) + phase + "_max.csv",infos, delimiter=',');

                
                    plt.subplot(1,nLayers,l+1)
                    if phaseIndex==1:
                        plt.plot(np.transpose(infos), linestyle='--', color='k');
                    elif phaseIndex==2:
                        plt.plot(np.transpose(infos), linestyle='-', color='k');
                    plt.ylim([-0.05, np.log2(nObj)+0.05]);
                    plt.xlim([0, nExcitCells])
                    plt.title("Layer " + str(l));
                    plt.ylabel("Information [bit]");
                    plt.xlabel("Cell Rank");
                    plt.hold(True);
                        
                

#         plt.savefig(os.path.split(os.path.realpath(__file__))[0] +"/Results/"+experimentName+"/singleCellInfo_bin"+str(nBins)+".png");
#         plt.savefig(os.path.split(os.path.realpath(__file__))[0] +"/Results/"+experimentName+"/singleCellInfo_bin"+str(nBins)+".eps");
            phaseIndex+=1;
        
        if showImage:
            plt.show();
        if saveImage:
            if (plotAllSingleCellInfo):
                fig.savefig("../output/"+experimentName+"/SingleCellInfo_ALL.png");
#                 fig.savefig("../output/SingleCellInfo_ALL.eps");
            else:
                  fig.savefig("../output/"+experimentName+"/SingleCellInfo_MAX.png");
#                   fig.savefig("../output/SingleCellInfo_MAX.eps");              
    
            print("figure SingleCellInfo.png is exported in Results") 
        
        plt.close();

File: MatLabPG/infoAnalysis.m
function IRs = infoAnalysis(inputMatrix,trained,nBins)
    plotAllSingleCellInfo = 0;
%     hold off;
%    num_bins =  2;%num_transforms;   %can be adjusted
    weightedAnalysis = 1;%exclude the selectivity by not responding to a particular stimulus


    num_transforms = size(inputMatrix,2);
    num_stimulus = size(inputMatrix,1);
    num_PGs = size(inputMatrix,3);

    showAllIRs = zeros(num_stimulus,num_PGs);



    disp('** Data loading **');

    sumPerBin = zeros(num_PGs,nBins);
    sumPerObj = num_transforms;
    sumPerCell = num_transforms*num_stimulus;
    IRs = zeros(num_PGs,num_stimulus);   %I(R,s) single cell information
    IRs_weighted = zeros(num_PGs,num_stimulus);   %I(R,s) single cell information

    binMatrix = zeros(num_PGs, num_stimulus, nBins); %number of times when fr is classified into a specific bin within a specific objects's transformations

    
    if max(inputMatrix(:))>0.001
        FR_tmp = inputMatrix./max(inputMatrix(:));
    end

    
    for obj = 1:num_stimulus;
        disp([num2str(obj) '/' num2str(num_stimulus)]);
        for trans = 1:num_transforms;
            for i = 1:num_PGs;
                bin = min(floor(FR_tmp(obj,trans,i))*nBins,nBins-1)+1;
                binMatrix(i,obj,bin)=binMatrix(i,obj,1)+1;
            end
        end
    end


    disp('DONE');
    disp(['** single-cell information analysis **']);
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % single-cell information analysis      %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    % Loop through all cells to calculate single cell information
    for i=1:num_PGs
        % For each cell, count the number of transforms per bin
        for bin=1:nBins
            sumPerBin(i,bin)=sum(binMatrix(i,:,bin));
        end

        % Calculate the information for cell_x cell_y per stimulus
        for obj=1:num_stimulus
            for bin=1:nBins
                Pr = sumPerBin(i,bin)/sumPerCell;
                Prs = binMatrix(i,obj,bin)/sumPerObj;
                if(Pr~=0&&Prs~=0&&Pr<Prs)
                    IRs(i,obj)=IRs(i,obj)+(Prs*(log2(Prs/Pr)));%*((bin-1)/(num_bins-1)); %could be added to weight the degree of firing rates.
                    %IRs(row,col,object)=IRs(row,col,object)+(Prs*(log2(Prs/Pr)))*((bin-1)/(num_bins-1)); %could be added to weight the degree of firing rates.
                    IRs_weighted(i,obj)=IRs_weighted(i,obj)+(Prs*(log2(Prs/Pr)))*((bin-1)/(nBins-1)); %could be added to weight the degree of firing rates.
                end
            end
        end
    end

    if (weightedAnalysis==1)
        IRs = IRs_weighted;
    end

    % Order by information content, descending
    IRs_tmp = sort(max(IRs,[],2), 'descend');%find max IRs for each 
    

    for obj = 1:num_stimulus
            tmp = [1:num_PGs;reshape(IRs(:,obj),1,[])];
            sorted = transpose(sortrows(transpose(tmp),-2));
    end
    %tmpstore = performanceArray;

    figure(1)
    hold on;
    for no = 1:num_stimulus
        showAllIRs(no,:) = sort(IRs(:,no), 'descend');
    end
    
    
    if(trained)
        if(plotAllSingleCellInfo)
            plot(transpose(showAllIRs), 'k-','LineWidth',2);
        else
            plot(IRs_tmp,'k-','LineWidth',2);
        end
    else
        if(plotAllSingleCellInfo)
            plot(transpose(showAllIRs), 'k--','LineWidth',2);
        else
            plot(IRs_tmp,'k--','LineWidth',2);
        end
    end
    
    
    %axis([0 num_PGs/10 -0.1 log2(num_stimulus)+0.1]);
    nCellsWithNonZeroInfo = length(find(IRs>0));
    axis([1 nCellsWithNonZeroInfo*1.2 -0.1 log2(num_stimulus)+0.1]);
    title('Single Cell Information Analysis');
    ylabel('Information [bit]');
    xlabel('Cell Rank');
    hold on;


%     dlmwrite([experiment '/IRs' num2str(layer) '.csv'],IRsHist);
%     dlmwrite([experiment '/IRs' num2str(layer) 'All_s' num2str(session) '.csv'],showAllIRs);
%     if(plotAllSingleCellInfo)
%         saveas(fig,[experiment '/info_single_all' num2str(layer) '.png']);
%         saveas(fig,[experiment '/info_single_all' num2str(layer) '.fig']);
%     else
%         saveas(fig,[experiment '/info_single' num2str(layer) '.png']);
%         saveas(fig,[experiment '/info_single' num2str(layer) '.fig']);
%     end


end


